File: code/packages/csharp/paint-vm/PaintVm.cs
using System.Collections;
using System.Reflection;
using System.Runtime.CompilerServices;
using CodingAdventures.PaintInstructions;
using PixelBuffer = CodingAdventures.PixelContainer.PixelContainer;

namespace CodingAdventures.PaintVm;

public static class PaintVmPackage
{
    public const string VERSION = "0.1.0";
}

public sealed class UnknownInstructionError(string kind) : Exception($"No handler registered for instruction kind: '{kind}'")
{
    public string Kind { get; } = kind;
}

public sealed class DuplicateHandlerError(string kind) : Exception($"Handler already registered for instruction kind: '{kind}'")
{
    public string Kind { get; } = kind;
}

public sealed class ExportNotSupportedError(string backendName)
    : Exception($"export() is not supported by the {backendName} backend. Use a backend that supports pixel readback.")
{
}

public sealed class NullContextError() : Exception("execute() and patch() require a non-null context")
{
}

public delegate void PaintHandler<TContext>(PaintInstructionBase instruction, TContext context, PaintVM<TContext> vm);

public sealed record ExportOptions
{
    public double Scale { get; init; } = 1.0;

    public int Channels { get; init; } = 4;

    public int BitDepth { get; init; } = 8;

    public string ColorSpace { get; init; } = "srgb";
}

public sealed record PatchCallbacks
{
    public Action<PaintInstructionBase>? OnDelete { get; init; }

    public Action<PaintInstructionBase, int>? OnInsert { get; init; }

    public Action<PaintInstructionBase, PaintInstructionBase>? OnUpdate { get; init; }
}

internal readonly struct ReferencePair(object left, object right)
{
    public object Left { get; } = left;

    public object Right { get; } = right;
}

internal sealed class ReferencePairComparer : IEqualityComparer<ReferencePair>
{
    public static ReferencePairComparer Instance { get; } = new();

    public bool Equals(ReferencePair x, ReferencePair y) =>
        ReferenceEquals(x.Left, y.Left) &&
        ReferenceEquals(x.Right, y.Right);

    public int GetHashCode(ReferencePair pair) =>
        HashCode.Combine(
            RuntimeHelpers.GetHashCode(pair.Left),
            RuntimeHelpers.GetHashCode(pair.Right));
}

/// <summary>
/// PaintVM is the dispatch-table execution engine for the paint IR. Backends
/// provide the clear function and the per-kind handlers; the VM supplies the
/// routing, diffing, and export contract.
/// </summary>
public sealed class PaintVM<TContext>
{
    private readonly Dictionary<string, PaintHandler<TContext>> _table = [];
    private readonly Action<TContext, string, double, double> _clear;
    private readonly Func<PaintScene, PaintVM<TContext>, ExportOptions, PixelBuffer>? _export;

    public PaintVM(
        Action<TContext, string, double, double> clear,
        Func<PaintScene, PaintVM<TContext>, ExportOptions, PixelBuffer>? export = null)
    {
        _clear = clear ?? throw new ArgumentNullException(nameof(clear));
        _export = export;
    }

    public void Register(string kind, PaintHandler<TContext> handler)
    {
        ArgumentException.ThrowIfNullOrEmpty(kind);
        ArgumentNullException.ThrowIfNull(handler);

        if (_table.ContainsKey(kind))
        {
            throw new DuplicateHandlerError(kind);
        }

        _table[kind] = handler;
    }

    public void Dispatch(PaintInstructionBase instruction, TContext context)
    {
        ArgumentNullException.ThrowIfNull(instruction);

        if (!_table.TryGetValue(instruction.Kind, out var handler) &&
            !_table.TryGetValue("*", out handler))
        {
            throw new UnknownInstructionError(instruction.Kind);
        }

        handler(instruction, context, this);
    }

    public void Execute(PaintScene scene, TContext context)
    {
        ArgumentNullException.ThrowIfNull(scene);
        EnsureContext(context);

        _clear(context, scene.Background, scene.Width, scene.Height);
        foreach (var instruction in scene.Instructions)
        {
            Dispatch(instruction, context);
        }
    }

    public void Patch(PaintScene oldScene, PaintScene newScene, TContext context, PatchCallbacks? callbacks = null)
    {
        ArgumentNullException.ThrowIfNull(oldScene);
        ArgumentNullException.ThrowIfNull(newScene);
        EnsureContext(context);

        if (callbacks is null)
        {
            Execute(newScene, context);
            return;
        }

        var oldById = oldScene.Instructions.Where(instruction => instruction.Id is not null)
            .ToDictionary(instruction => instruction.Id!, instruction => instruction);
        var newById = newScene.Instructions.Where(instruction => instruction.Id is not null)
            .ToDictionary(instruction => instruction.Id!, instruction => instruction);

        foreach (var (id, instruction) in oldById)
        {
            if (!newById.ContainsKey(id))
            {
                callbacks.OnDelete?.Invoke(instruction);
            }
        }

        for (var index = 0; index < newScene.Instructions.Count; index++)
        {
            var nextInstruction = newScene.Instructions[index];

            if (nextInstruction.Id is not null && oldById.TryGetValue(nextInstruction.Id, out var oldInstruction))
            {
                if (!DeepEqual(nextInstruction, oldInstruction))
                {
                    callbacks.OnUpdate?.Invoke(oldInstruction, nextInstruction);
                }

                continue;
            }

            if (index < oldScene.Instructions.Count)
            {
                var positionalOld = oldScene.Instructions[index];
                if (!DeepEqual(nextInstruction, positionalOld))
                {
                    callbacks.OnUpdate?.Invoke(positionalOld, nextInstruction);
                }
            }
            else
            {
                callbacks.OnInsert?.Invoke(nextInstruction, index);
            }
        }
    }

    public PixelBuffer Export(PaintScene scene, ExportOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(scene);

        if (_export is null)
        {
            throw new ExportNotSupportedError("this");
        }

        return _export(scene, this, options ?? new ExportOptions());
    }

    public IReadOnlyList<string> RegisteredKinds() => _table.Keys.Order().ToArray();

    public static bool DeepEqual(object? left, object? right) =>
        DeepEqual(left, right, new HashSet<ReferencePair>(ReferencePairComparer.Instance));

    private static bool DeepEqual(object? left, object? right, HashSet<ReferencePair> visited)
    {
        if (ReferenceEquals(left, right))
        {
            return true;
        }

        if (left is null || right is null)
        {
            return false;
        }

        if (left.GetType() != right.GetType())
        {
            return false;
        }

        var type = left.GetType();

        if (left is string || type.IsPrimitive || left is decimal || type.IsEnum)
        {
            return Equals(left, right);
        }

        if (!type.IsValueType && !visited.Add(new ReferencePair(left, right)))
        {
            return true;
        }

        if (left is IDictionary leftDictionary && right is IDictionary rightDictionary)
        {
            if (leftDictionary.Count != rightDictionary.Count)
            {
                return false;
            }

            foreach (var key in leftDictionary.Keys)
            {
                if (!rightDictionary.Contains(key) || !DeepEqual(leftDictionary[key], rightDictionary[key], visited))
                {
                    return false;
                }
            }

            return true;
        }

        if (left is IEnumerable leftEnumerable && right is IEnumerable rightEnumerable)
        {
            var leftItems = leftEnumerable.Cast<object?>().ToArray();
            var rightItems = rightEnumerable.Cast<object?>().ToArray();
            if (leftItems.Length != rightItems.Length)
            {
                return false;
            }

            for (var index = 0; index < leftItems.Length; index++)
            {
                if (!DeepEqual(leftItems[index], rightItems[index], visited))
                {
                    return false;
                }
            }

            return true;
        }

        foreach (var property in left.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance).Where(property => property.CanRead))
        {
            if (!DeepEqual(property.GetValue(left), property.GetValue(right), visited))
            {
                return false;
            }
        }

        return true;
    }

    private static void EnsureContext(TContext context)
    {
        if (context is null)
        {
            throw new NullContextError();
        }
    }
}

File: code/packages/fsharp/paint-vm/PaintVm.fs
namespace CodingAdventures.PaintVm

open System
open System.Collections.Generic
open CodingAdventures.PaintInstructions
open CodingAdventures.PixelContainer

[<RequireQualifiedAccess>]
module PaintVmPackage =
    [<Literal>]
    let VERSION = "0.1.0"

type UnknownInstructionError(kind: string) =
    inherit Exception(sprintf "No handler registered for instruction kind: '%s'" kind)
    member _.Kind = kind

type DuplicateHandlerError(kind: string) =
    inherit Exception(sprintf "Handler already registered for instruction kind: '%s'" kind)
    member _.Kind = kind

type ExportNotSupportedError(backendName: string) =
    inherit Exception(sprintf "export() is not supported by the %s backend. Use a backend that supports pixel readback." backendName)

type NullContextError() =
    inherit Exception("execute() and patch() require a non-null context")

type ExportOptions =
    {
        Scale: float
        Channels: int
        BitDepth: int
        ColorSpace: string
    }

type PatchCallbacks =
    {
        OnDelete: (PaintInstruction -> unit) option
        OnInsert: (PaintInstruction -> int -> unit) option
        OnUpdate: (PaintInstruction -> PaintInstruction -> unit) option
    }

type PaintHandler<'Context> = PaintInstruction -> 'Context -> PaintVM<'Context> -> unit

and PaintVM<'Context>
    (
        clear: 'Context -> string -> float -> float -> unit,
        ?exporter: PaintScene -> PaintVM<'Context> -> ExportOptions -> PixelContainer
    ) =

    let table = Dictionary<string, PaintHandler<'Context>>()
    let exporter = exporter

    let ensureContext (context: 'Context) =
        if obj.ReferenceEquals(context, null) then
            raise (NullContextError())

    member _.Register(kind: string, handler: PaintHandler<'Context>) =
        if String.IsNullOrWhiteSpace(kind) then
            invalidArg "kind" "kind must not be empty"

        if isNull (box handler) then
            nullArg "handler"

        if table.ContainsKey(kind) then
            raise (DuplicateHandlerError(kind))

        table[kind] <- handler

    member this.Dispatch(instruction: PaintInstruction, context: 'Context) =
        let mutable handler = Unchecked.defaultof<PaintHandler<'Context>>

        let foundSpecific = table.TryGetValue(instruction.Kind, &handler)
        let mutable foundWildcard = false

        if not foundSpecific then
            let mutable wildcardHandler = Unchecked.defaultof<PaintHandler<'Context>>

            if table.TryGetValue("*", &wildcardHandler) then
                handler <- wildcardHandler
                foundWildcard <- true

        if foundSpecific || foundWildcard then
            handler instruction context this
        else
            raise (UnknownInstructionError(instruction.Kind))

    member this.Execute(scene: PaintScene, context: 'Context) =
        ensureContext context
        clear context scene.Background scene.Width scene.Height
        scene.Instructions |> List.iter (fun instruction -> this.Dispatch(instruction, context))

    member this.Patch(oldScene: PaintScene, newScene: PaintScene, context: 'Context, ?callbacks: PatchCallbacks) =
        ensureContext context

        match callbacks with
        | None -> this.Execute(newScene, context)
        | Some callbacks ->
            let oldById =
                oldScene.Instructions
                |> List.choose (fun instruction -> instruction.Id |> Option.map (fun id -> id, instruction))
                |> Map.ofList

            let newById =
                newScene.Instructions
                |> List.choose (fun instruction -> instruction.Id |> Option.map (fun id -> id, instruction))
                |> Map.ofList

            for KeyValue(id, instruction) in oldById do
                if not (newById.ContainsKey(id)) then
                    callbacks.OnDelete |> Option.iter (fun onDelete -> onDelete instruction)

            for index, nextInstruction in newScene.Instructions |> List.indexed do
                match nextInstruction.Id with
                | Some id when oldById.ContainsKey(id) ->
                    let oldInstruction = oldById[id]

                    if not (PaintVM<'Context>.DeepEqual(box nextInstruction, box oldInstruction)) then
                        callbacks.OnUpdate |> Option.iter (fun onUpdate -> onUpdate oldInstruction nextInstruction)
                | _ ->
                    if index < oldScene.Instructions.Length then
                        let positionalOld = oldScene.Instructions[index]

                        if not (PaintVM<'Context>.DeepEqual(box nextInstruction, box positionalOld)) then
                            callbacks.OnUpdate |> Option.iter (fun onUpdate -> onUpdate positionalOld nextInstruction)
                    else
                        callbacks.OnInsert |> Option.iter (fun onInsert -> onInsert nextInstruction index)

    member this.Export(scene: PaintScene, ?options: ExportOptions) =
        match exporter with
        | Some export ->
            export scene
                this
                (defaultArg options
                    {
                        Scale = 1.0
                        Channels = 4
                        BitDepth = 8
                        ColorSpace = "srgb"
                    })
        | None -> raise (ExportNotSupportedError("this"))

    member _.RegisteredKinds() =
        table.Keys |> Seq.sort |> Seq.toArray

    static member DeepEqual(left: obj, right: obj) =
        if obj.ReferenceEquals(left, right) then
            true
        elif isNull left || isNull right then
            false
        elif left.GetType() <> right.GetType() then
            false
        else
            left.Equals(right)
